### DISCUSSION TITLE
How spack deals with dependencies?

### DISCUSSION THREAD

**Discussion by SJ2050cn** [spack/spack#31498] [Q&A]
Everytime we run command `spack info xxx`, it will show like follows:
```
AutotoolsPackage:   mpileaks

Description:
    Tool to detect and report MPI objects like MPI_Requests and
    MPI_Datatypes.

Homepage: https://github.com/LLNL/mpileaks

Preferred version:
    1.0    https://github.com/LLNL/mpileaks/releases/download/v1.0/mpileaks-1.0.tar.gz

Safe versions:
    1.0    https://github.com/LLNL/mpileaks/releases/download/v1.0/mpileaks-1.0.tar.gz

Deprecated versions:
    None

Variants:
    Name [Default]    When    Allowed values          Description
    ==============    ====    ====================    ==========================================
....

Build Dependencies:
    adept-utils  callpath  gnuconfig  mpi

Link Dependencies:
    adept-utils  callpath  mpi

Run Dependencies:
    None
```
**My Problem is how spack knows which dependency is for build, link or run.** I have read some package.py, but I still have no idea because I can't find any explicitly specification in package.py. Sincerely thanks for your answer in advance.

**Comment by climbfuji** [comment]
spack has defaults for dependencies, i.e. if you don't specify it in `package.py` it assumes build and link:

https://spack.readthedocs.io/en/latest/spack.html?highlight=dependency#spack.dependency.default_deptype

**Reply by SJ2050cn** [reply]
> spack has defaults for dependencies, i.e. if you don't specify it in `package.py` it assumes build and link:
> 
> https://spack.readthedocs.io/en/latest/spack.html?highlight=dependency#spack.dependency.default_deptype

Can you explain `spack.dependency.default_deptype` basic working principle? The document is too short. Spack assums build and link according to which information. Thank you.
